docs/source/grammar.rst

```diff
@@ -0,0 +1,240 @@
+
+
+Escaped characters
+^^^^^^^^^^^^^^^^^^
+
+Some characters need to be escaped, with varying escaping requirements depending on the situation.
+In general, however, you can use the following rule of thumb:
+*you only need to escape characters that otherwise have a special meaning in the current context*.
+
+.. _escaping-in-sring-interpolations:
+
+Escaping in string interpolations
++++++++++++++++++++++++++++++++++
+
+In order to define fields whose value is an interpolation-like string, interpolations can be escaped with ``\${``.
+For instance:
+
+.. doctest::
+
+    >>> c = OmegaConf.create({"path": r"\${dir}", "dir": "tmp"})
+    >>> print(c.path)  # does *not* interpolate into the `dir` node
+    ${dir}
+
+Since the ``\`` character has a special meaning in front of ``${``,
+it must itself be escaped with ``\\`` if you actually intend to follow it with an interpolation:
+
+.. doctest::
+
+    >>> c = OmegaConf.create({"path": r"C:\\${dir}", "dir": "tmp"})
+    >>> print(c.path)  # *does* interpolate into the `dir` node
+    C:\tmp
+
+Finally, since the ``\`` character has no special meaning unless followed by ``${``,
+it does *not* need to be escaped anywhere else:
+
+.. doctest::
+
+    >>> c = OmegaConf.create({"path": r"C:\foo_${dir}", "dir": "tmp"})
+    >>> print(c.path)  # a single \ is preserved...
+    C:\foo_tmp
+    >>> c = OmegaConf.create({"path": r"C:\\foo_${dir}", "dir": "tmp"})
+    >>> print(c.path)  # ... and multiple \\ too (no escape sequence)
+    C:\\foo_tmp
+
+Escaping in unquoted strings
+++++++++++++++++++++++++++++
+
+Unquoted strings can be found in a number of contexts, including dictionary keys/values,
+list elements, etc. As a result, the  escape sequences are used for some
+special characters
+(``\\``, ``\[``, ``\]``, ``\{``, ``\}``, ``\(``, ``\)``, ``\:``, ``\=``, ``\,``),
+for instance:
+
+    - ``C\:\\$\{dir\}`` resolves to the string ``"C:\${dir}"``
+    - ``\[a\, b\, c\]`` resolves to the string ``"[a, b, c]"``
+
+In addition, leading and trailing whitespaces must be escaped in unquoted strings
+if we do not want them to be stripped (while inner whitespaces are always preserved):
+
+.. doctest::
+
+    >>> c = OmegaConf.create({"esc": r"${oc.decode: \ hi u \  }"})
+    >>> c.esc  # one leading whitespace and two trailing ones
+    ' hi u  '
+    >>> # Tabs are handled similarly (NB: r-strings can't be used below)
+    >>> c = OmegaConf.create({"esc": "${oc.decode:\t\\\thi u\t\\\t\t}"})
+    >>> c.esc  # one leading tab and two trailing ones
+    '\thi u\t\t'
+
+Escaping in unquoted strings can lead to hard-to-read expressions, and it is recommended
+to switch to quoted strings instead of relying heavily on the above escape sequences.
+
+Escaping in quoted strings
+++++++++++++++++++++++++++
+
+As can be seen from the definition of the ``quotedValue`` parser rule above, a quoted string
+is essentially a string interpolation surrounded by quotes.
```

Comment:
Good point, this is a consequence of some over-simplifications I made. I tried to make it clearer in bb77fc1 => let me know if that's better.

I changed "string interpolation" into "interpolation string" because "string interpolation" actually means something very specific: the concatenation of at least one interpolation with either some regular string or another interpolation (which always results in a string result, e.g., `https://${host}:${port}`). But the ``text`` parser rule is also used to parse single interpolations like ``${foo}`` (that aren't string interpolations).

Reply:
Looks good!